cache/clock_cache.cc

```diff
@@ -503,23 +512,27 @@ Status ClockCacheShard::Insert(const Slice& key, uint32_t hash, void* value,
       }
     }
   } else {
-    ClockHandle* h;
-    if (occupancy_local + 1 > table_.GetOccupancyLimit()) {
+    ClockHandle* h = nullptr;
+    if (handle != nullptr && occupancy_local + 1 > table_.GetOccupancyLimit()) {
       // Even if the user wishes to overload the cache, we can't insert into
       // the hash table. Instead, we dynamically allocate a new handle.
-      h = new ClockHandle();
-      *h = tmp;
-      h->SetDetached();
-      h->TryExternalRef();
-      detached_usage_ += h->total_charge;
+      h = DetachedInsert(&tmp);
       // TODO: Return special status?
     } else {
       // Insert into the cache. Note that the cache might get larger than its
       // capacity if not enough space was freed up.
       autovector<ClockHandle> deleted;
       h = table_.Insert(&tmp, &deleted, handle != nullptr);
-      assert(h != nullptr);  // The occupancy is way below the table size, so
-                             // this insertion should never fail.
+      if (h == nullptr) {
+        // The table is full. This can happen when many threads simultaneously
+        // attempt an insert, and the table is operating close to full capacity.
+        if (handle != nullptr) {
+          h = DetachedInsert(&tmp);
+        } else {
+          s = Status::MemoryLimit(
```

Comment:
I'm pretty sure LRUCache uses OK for this kind of case, and therefore we should also. https://github.com/facebook/rocksdb/blob/8b2d4292514c7ad82b801a97dc7df02cf760f015/cache/lru_cache.cc#L327